petastorm/spark/spark_dataset_converter.py

```diff
@@ -442,16 +451,61 @@ def _cache_df_or_retrieve_cache_data_url(df, parent_cache_dir_url,
         for meta in _cache_df_meta_list:
             if meta.row_group_size == parquet_row_group_size_bytes and \
                     meta.compression_codec == compression_codec and \
-                    meta.df_plan.sameResult(df_plan):
+                    meta.df_plan.sameResult(df_plan) and \
+                    meta.dtype == dtype:
                 return meta.cache_dir_url
         # do not find cached dataframe, start materializing.
         cached_df_meta = CachedDataFrameMeta.create_cached_dataframe(
             df, parent_cache_dir_url, parquet_row_group_size_bytes,
-            compression_codec)
+            compression_codec, dtype)
         _cache_df_meta_list.append(cached_df_meta)
         return cached_df_meta.cache_dir_url
 
 
+def _convert_precision(df, dtype):
+    if dtype != "float32" and dtype != "float64":
+        raise ValueError("dtype {} is not supported. \
+            Use 'float32' or float64".format(dtype))
+
+    source_type, target_type = (DoubleType, FloatType) \
+        if dtype == "float32" else (FloatType, DoubleType)
+
+    for struct_field in df.schema:
+        col_name = struct_field.name
+        if isinstance(struct_field.dataType, source_type):
+            df = df.withColumn(col_name, df[col_name].cast(target_type()))
+        elif isinstance(struct_field.dataType, ArrayType) and \
+                isinstance(struct_field.dataType.elementType, source_type):
+            df = df.withColumn(col_name, df[col_name].cast(ArrayType(target_type())))
+    return df
+
+
+def _convert_vector(df, dtype):
+    from pyspark.ml.linalg import Vector
+    from pyspark.ml.linalg import VectorUDT
+    from pyspark.mllib.linalg import Vector as OldVector
+    from pyspark.mllib.linalg import VectorUDT as OldVectorUDT
+
+    found_vector_column = False
+    for struct_field in df.schema:
+        col_name = struct_field.name
+        if struct_field.dataType == Vector() or \
+                struct_field.dataType == OldVector() or \
+                type(struct_field.dataType) == VectorUDT or \
+                type(struct_field.dataType) == OldVectorUDT:
+            if not found_vector_column:
+                import pyspark
+                if LooseVersion(pyspark.__version__) < LooseVersion('3.0'):
+                    raise ValueError("Vector columns are not supported for pyspark<3.0.0.")
+                # pylint: disable=import-error
+                from pyspark.ml.functions import vector_to_array
+                # pylint: enable=import-error
+                found_vector_column = True
```

Comment:
The logic of `found_vector_column` looks weird. What about move it to the top of this module, like:

```
if LooseVersion(pyspark.__version__) < LooseVersion('3.0'):
   vector_to_array = lambda col, dtype='float32': raise RuntimeError("only supported in spark>=3.0")
else:
  from pyspark.ml.functions import vector_to_array
```